### PR TITLE
L2-476 and L2-470: Improve Form Input Validation Error when adding a Hotkey

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

User receives immediate feedback when entering a wrong hotkey.

# Changes

* Extend Input element to show an error message.
* Extend Input element to listen for blur event.
* Validate address entered on blur in AddHotkeyModal.

# Tests

* Extend Input tests with new functionality.
* Extend AddHotkeyModal tests with validation scenarios.

### CODE REVIEW

**Comment by lmuntanerDfinity** [comment]
UX video:

https://user-images.githubusercontent.com/98735375/165939001-ce4f4a23-fa07-4b28-855f-6ba7ed18e8cd.mov

Screenshots:

Screenshot 2022-04-29 at 13 44 23
Screenshot 2022-04-29 at 13 49 23

**Comment by lmuntanerDfinity** [comment]
@mstrasinskis @peterpeterparker please review

**Comment by peterpeterparker** [comment]
as discussed this first version leads to layout shift but the input itsel should not move in my opinion. with layout shift personally I understood the display of the error moves whats below. can you please improve this?

Capture d’écran 2022-05-02 à 07 07 56

File: frontend/svelte/src/lib/components/ui/Input.svelte [hunk 1]
```diff
@@ -28,7 +30,11 @@
   $: placeholder = translate({ labelKey: placeholderLabelKey });
 </script>
 
-<div class={`input-block ${theme}`} class:disabled>
+<div
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/Input.svelte:33]: - do we really need that `div`? if we had it then all inputs will be rendered with an extra div in the DOM

- if we really need it, then should we create rather create a new component `InputWithError` or whatever because it isn't just an input anymore?

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte [hunk 2]
```diff
@@ -13,23 +13,46 @@
   export let neuronId: NeuronId;
 
   let address: string = "";
-  let showAddressNotValid: boolean = false;
+  let validPrincipal: Principal | undefined;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:16]: shouldn't this be a utility function rather than being implemented within the component?

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte [hunk 3]
```diff
@@ -13,23 +13,46 @@
...
-      showAddressNotValid = true;
+      prevValidatedAddress = address;
+      Principal.fromText(address);
+      showError = false;
+    } catch (_) {
+      showError = true;
     }
-    if (principal !== undefined) {
+  };
+  // Hide error when user is changing the value.
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:44]: from the comment I excepted

```
$: showError = false;
```

can you rephrase the comment?

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte [hunk 4]
```diff
@@ -13,23 +13,46 @@
...
+    }
+  }
+  let showError: boolean = false;
   let loading: boolean = false;
 
-  const dispatcher = createEventDispatcher();
-  const add = async () => {
-    let principal: Principal | undefined = undefined;
+  // Used to hide error when address changes.
+  let prevValidatedAddress: string | undefined;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:28]: do we really want to not hide the error when user type?

to some extension we hide the error but then on blur the error can popup again, so is it really useful?

should we validate the adress on input with a debounce aswell?

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
Couple of comments + layout shift

File: frontend/svelte/src/lib/components/ui/Input.svelte (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/Input.svelte:33]: ok

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:16]: Sure

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte (hunk 3 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:44]: ok

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte (hunk 4 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:28]: I followed what Mischa [wrote](https://www.notion.so/Form-validation-messages-on-dark-background-43afe1a80b0b402c945855b5cdd9eb16#c0ffa14328fb4579a9ef26ac0bf3faec):

> - Validate on blur (add validation message/error on blur)
> - Remove validation message/error on change

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker ready for another review

I could have used the same `Input`, but I liked better your idea of a new Input. If we don't want layout shift, we need to make sure the space is there even if we don't show the error. Which I did using a margin-bottom.

Yet, since not all inputs want or need this space, it made more sense to have the `InputWithError` component for those place where we can assume this extra bottom space.

File: frontend/svelte/src/lib/components/ui/Input.svelte [hunk 5]
```diff
@@ -28,8 +30,9 @@
   $: placeholder = translate({ labelKey: placeholderLabelKey });
 </script>
 
-<div class={`input-block ${theme}`} class:disabled>
+<div class={`input-block ${theme}`} class:disabled class:error>
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/Input.svelte:33]: this should move to `InputWithError` no?

File: frontend/svelte/src/lib/components/ui/InputWithError.svelte [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  export let errorMessage: string | undefined = undefined;
+  // https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete
+  export let autocomplete: "off" | "on" | undefined = undefined;
+  export let value: string | number | undefined = undefined;
+  export let placeholderLabelKey: string;
+  export let theme: "dark" | "light" = "light";
+</script>
+
+<div class="wrapper">
+  <Input
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/InputWithError.svelte:23]: `{...$$props}` or that does not work here?

**Review by peterpeterparker — COMMENTED** [review]
two comments

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte (hunk 4 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:23]: Would just `$: value, showError = false` work then?

File: frontend/svelte/src/lib/components/ui/InputWithError.svelte (hunk 6 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/InputWithError.svelte:23]: I'd rather be explicit.

File: frontend/svelte/src/lib/components/ui/Input.svelte (hunk 5 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/Input.svelte:33]: The border of the input is set in this component. I guess I could do it without a prop, but I think a prop is the most readable option.

File: frontend/svelte/src/lib/components/ui/InputWithError.svelte (hunk 6 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/ui/InputWithError.svelte:23]: oki doki

File: frontend/svelte/src/lib/components/ui/Input.svelte (hunk 5 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/ui/Input.svelte:33]: forget my question, thanks for the answer

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker, I forgot to push a new video with the UX:

https://user-images.githubusercontent.com/98735375/166227161-d6d7b3e7-4b4a-45c1-9937-7f7788f32276.mov

File: frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte (hunk 4 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/AddHotkeyModal.svelte:23]: It does! I love these Svelte tricks, nice one!

**Review by peterpeterparker — APPROVED** [review]
thanks
